File: intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/beans/PluginBeanExtractor.java
package com.jetbrains.plugin.structure.intellij.beans;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import org.jdom2.Document;
import org.jdom2.Element;
import org.jdom2.Namespace;
import org.jdom2.transform.JDOMSource;
import org.jetbrains.annotations.NotNull;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;

public class PluginBeanExtractor {

  @NotNull
  public static PluginBean extractPluginBean(Document document) throws JAXBException {
    JAXBContext jc = JAXBContext.newInstance(PluginBean.class);
    Unmarshaller unmarshaller = jc.createUnmarshaller();

    Element rootElement = document.getRootElement();
    PluginBean bean = (PluginBean) unmarshaller.unmarshal(new JDOMSource(document));
    bean.extensions = extractExtensions(rootElement);
    return bean;
  }

  private static Multimap<String, Element> extractExtensions(Element rootElement) {
    Multimap<String, Element> extensions = ArrayListMultimap.create();
    for (Element extensionsRoot : rootElement.getChildren("extensions")) {
      for (Element element : extensionsRoot.getChildren()) {
        extensions.put(extractEPName(element), element);
      }
    }
    return extensions;
  }

  private static String extractEPName(final Element extensionElement) {
    String epName = extensionElement.getAttributeValue("point");

    if (epName == null) {
      final Element parentElement = extensionElement.getParentElement();
      final String ns = parentElement != null ? parentElement.getAttributeValue("defaultExtensionNs") : null;

      if (ns != null) {
        epName = ns + '.' + extensionElement.getName();
      } else {
        Namespace namespace = extensionElement.getNamespace();
        epName = namespace.getURI() + '.' + extensionElement.getName();
      }
    }
    return epName;
  }

}

File: intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/beans/PluginBeanExtractor.kt
package com.jetbrains.plugin.structure.intellij.beans

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import org.jdom2.Document
import org.jdom2.Element
import org.jdom2.transform.JDOMSource
import javax.xml.bind.JAXBContext
import javax.xml.bind.JAXBException

object PluginBeanExtractor {

  private val jaxbContext by lazy {
    JAXBContext.newInstance(PluginBean::class.java)
  }

  @Throws(JAXBException::class)
  fun extractPluginBean(document: Document): PluginBean {
    val unmarshaller = jaxbContext.createUnmarshaller()

    val rootElement = document.rootElement
    val bean = unmarshaller.unmarshal(JDOMSource(document)) as PluginBean
    bean.extensions = extractExtensions(rootElement)
    return bean
  }

  private fun extractExtensions(rootElement: Element): Multimap<String, Element> {
    val extensions = ArrayListMultimap.create<String, Element>()
    for (extensionsRoot in rootElement.getChildren("extensions")) {
      for (element in extensionsRoot.children) {
        extensions.put(extractEPName(element), element)
      }
    }
    return extensions
  }

  private fun extractEPName(extensionElement: Element): String {
    var epName: String? = extensionElement.getAttributeValue("point")

    if (epName == null) {
      val parentElement = extensionElement.parentElement
      val ns = parentElement?.getAttributeValue("defaultExtensionNs")

      if (ns != null) {
        epName = ns + '.'.toString() + extensionElement.name
      } else {
        val namespace = extensionElement.namespace
        epName = namespace.uri + '.'.toString() + extensionElement.name
      }
    }
    return epName
  }

}
